### PR TITLE
Watch Only Specified mode

### PR DESCRIPTION
**Pull request by GhatSmith**
Watch Only Specified mode : file watchers cleared and fast script reload disabled when starting a new editor session + enabling fast script reload only when file watcher is not empty
New MenuItems to add/remove file watcher from project window contextual menu

### CODE REVIEW

**Comment by GhatSmith** [comment]
@handzlikchris Things you may want to improve:

1. Move "Watch Only Specified" toggle to a better place (File Watchers menu maybe?). I've put it in the Enabled Features so it's easy to access.
2. Use an enum field "Watch Everything" / "Watch Only Specified" instead of a toggle (guess it would require something like EnumProjectEditorPreferenceDefinition).
3. Improve menu item naming
4. When you enable the Watch Only Specified mode, the "Enable auto hot-reload for changed files (in play mode)" toggle Is now disabled (because it's state is handled automatically, disabled when file watchers is empty). You would probably need to add an explanation message.
5. Set FileWatcherSetupEntriesChanged flag to true when file watchers are changed in the editor window (would require to expose a value change callback from ProjectEditorPreferenceDefinitionBase)
6. You may have false positive console warning ("Both auto hot reload and on-demand reload are disabled...") when Watch Only Specified is enabled

**Comment by handzlikchris** [comment]
Really cool - thanks! I'll check it out properly early next week

**Comment by handzlikchris** [comment]
got sidetracked with holidays! Still not lost on my todo

**Comment by handzlikchris** [comment]
changed base for edits / testing

File: Assets/Scripts/Editor/FastScriptReloadManager.cs [hunk 1]
```diff
@@ -181,7 +192,125 @@ static FastScriptReloadManager()
                 ClearFileWatchers();
             }
         }
-        
+
+        private const int BaseMenuItemPriority_FileWatcher = 0;
+        [MenuItem("Assets/Fast Script Reload/Watch Files", true, BaseMenuItemPriority_FileWatcher + 1)]
+        public static bool ToggleSelectionFileWatchersSetupValidation()
+        {
+            Menu.SetChecked("Assets/Fast Script Reload/Watch Files", false);
```
Comment (handzlikchris) [inline Assets/Scripts/Editor/FastScriptReloadManager.cs:200]: that's cool I didn't knew about that!

**Comment by handzlikchris** [comment]
That's now all merged, that's really cool feature.

Good idea to hook that up to existing FileWatchers functionality

Thanks!

**Comment by digiwombat** [comment]
> Packages\com.handzlikchris.fastscriptreload@003de5f0e1\Scripts\Editor\NewFields\NewFieldsRendererDefaultEditorPatch.cs(34,141): error CS0103: The name 'DrawOdinInspector' does not exist in the current context
This introduced the above error when installing via git.
Odin Inspector is at 3.1.14.1 (latest)

**Comment by builder-main** [comment]
Gets the above issue as well, it does not resolve the nameof. Dirty workaround as of now replace by string "DrawOdinInspector"
 var renderAdditionalFieldsDrawOdinInspectorPostfix = AccessTools.Method(typeof(NewFieldsRendererDefaultEditorPatch), nameof(DrawOdinInspector));
Should prbly check for odin installation somehow to add the reference. Or just the plain string for now to get everyone along.

**Comment by builder-main** [comment]
Should it work for added functions ? 

```c#
   [Sirenix.OdinInspector.Button]
        [ShowInInspector]
        [DisableInEditorMode]
        private static void MyFunction()
        {


        }
```


```
System.ArgumentNullException: Value cannot be null.
Parameter name: method
  at HarmonyLib.HarmonyMethod..ctor (System.Reflection.MethodInfo method) [0x00010] in <5c856fb19ddb469d99459756a2e14341>:0 
  at FastScriptReload.Editor.NewFields.NewFieldsRendererDefaultEditorPatch..cctor () [0x000c1] in
```

**Comment by builder-main** [comment]
Had to fix it this way
```c#
#if ODIN_INSPECTOR
                // Odin Inspector support
                var renderAdditionalFieldsDrawOdinInspectorPostfix = AccessTools.Method(typeof(NewFieldsRendererDefaultEditorPatch), "DrawOdinInspector");
                var customOdinEditorRenderingMethod = AccessTools.Method("Sirenix.OdinInspector.Editor.OdinEditor:DrawOdinInspector");
                if(renderAdditionalFieldsDrawOdinInspectorPostfix == null) Debug.LogWarning("Unable to find method NewFieldsRendererDefaultEditorPatch : DrawOdinInspector");
                else harmony.Patch(customOdinEditorRenderingMethod, postfix: new HarmonyMethod(renderAdditionalFieldsDrawOdinInspectorPostfix));
#endif
```

**Comment by builder-main** [comment]
Sorry just noticed I'm posting on the Wrong PR.

**Comment by GhatSmith** [comment]
> > Packages\com.handzlikchris.fastscriptreload@[003de5f](https://github.com/handzlikchris/FastScriptReload/commit/003de5f0e10536193155bfa7311f3b9d438ba5ea)\Scripts\Editor\NewFields\NewFieldsRendererDefaultEditorPatch.cs(34,141): error CS0103: The name 'DrawOdinInspector' does not exist in the current context
> 
> This introduced the above error when installing via git. Odin Inspector is at 3.1.14.1 (latest)

@digiwombat  @builder-main @handzlikchris 
Sorry I only just noticed now that there was a problem with my pull request regarding Odin compatibility. This is easy to fix, [see details in the appropriate place](https://github.com/handzlikchris/FastScriptReload/issues/77#issuecomment-1786965409).

**Comment by handzlikchris** [comment]
updated code so compile error will be gone.

**Pull request merged**